### PR TITLE
Stream output files instead of writing them in-memory first 

### PR DESCRIPTION
**Pull request by Mikolaj-A-Kowalski**
This resolves Issue #38.  However, I cannot help but to think it is a bit filthy... that being said I am not sure how exactly to make it nicer. 

The problem with streaming to the output file directly is that it is convenient to have a 'backtrace' capability, when writing the printers for different output formats. This allows, for example, to remove trailing commas with relative ease (crucial for JSON). 

To 'have a cake and eat it' this implements a `delayedStream` which works a bit like a buffered output, but when the buffer is flushed it retains `MIN_SIZE=16` last characters [as opposed to printing everything to the disk]. 

What I don't like about the changes is the way the file ownership is handled. It is made to be a responsibility of the `outputFile` class, which opens a file in `init`and closes it in `finalisation` (which is also the destructor). However, the printing is implemented by the `asciiOutput_inter` which owns an instance of `delayedStream`. The appropriate output unit is set by a subroutine.  

The other thing I fear may be confusing is that if one does not supply `filename` to `outputFile` no file will be produced (this is required for unit testing). 

In short this PR may still require some iterations [in addition to squashing the fixups!]

### CODE REVIEW

File: UserInterface/fileOutput/asciiJSON_class.f90 [hunk 1]
```diff
@@ -71,51 +70,43 @@ subroutine init(self)
...
     end if
+    call self % append("}" // NEWLINE)
 
-    ! Close the file
-    call self % output % append("}")
+  end subroutine endFile
 
-    form = '(A' // numToChar(self % output % length()) // ')'
+  !!
+  !! Return approperiate extension for the file
```
Comment (valeriaRaffuzzi) [inline UserInterface/fileOutput/asciiJSON_class.f90:99]:
```suggestion
  !! Return appropriate extension for the file
```

File: UserInterface/fileOutput/asciiMATLAB_class.f90 [hunk 2]
```diff
@@ -90,33 +87,29 @@ subroutine init(self)
...
 
-    str = 'm'
+    ! Nothing to do
 
-  end function extension
+  end subroutine endFile
 
   !!
-  !! Print the output to the given unit
+  !! Return approperiate extension for the file
```
Comment (valeriaRaffuzzi) [inline UserInterface/fileOutput/asciiMATLAB_class.f90:102]:
```suggestion
  !! Return appropriate extension for the file
```

File: UserInterface/fileOutput/dummyPrinter_class.f90 [hunk 3]
```diff
@@ -23,6 +23,7 @@ module dummyPrinter_class
 
   contains
     procedure :: init
+    procedure :: endFile
     procedure :: extension
     procedure :: writeToFile
```
Comment (valeriaRaffuzzi) [inline UserInterface/fileOutput/dummyPrinter_class.f90:28]: I thought writeToFile was gone in the asciiOutput class?

File: UserInterface/fileOutput/outputFile_class.f90 [hunk 4]
```diff
@@ -210,17 +211,54 @@ module outputFile_class
...
     allocate( self % output, source = new_asciiOutput(self % type))
 
+    ! Open file if given
+    ! Otherwise make sure that the unit is not opened
+    if (present(filename)) then
+      ! Because GFotran hates implicit allocation
+      ! It causes buggy warnings to appear in compilation
+      allocate(self % outputFileName, source = trim(filename) // "." // self % output % extension())
+
+      open(newunit = self % outputUnit, &
```
Comment (valeriaRaffuzzi) [inline UserInterface/fileOutput/outputFile_class.f90:240]: It looks like 'error' is gone from this list.

Comment (valeriaRaffuzzi) [reply UserInterface/fileOutput/outputFile_class.f90:240]: same with errorMsg

File: UserInterface/fileOutput/outputFile_class.f90 [hunk 5]
```diff
@@ -233,58 +271,32 @@ subroutine init(self, type, fatalErrors)
...
-    ! Catch error
-    if (error /= 0 ) call fatalError(Here, errorMsg)
-
-    ! Write to file
-    call self % output % writeToFile(unitNum)
-
-    ! Close file
-    close(unitNum)
+  !! NOTE:
+  !!   This subroutine WILL NOT be called if the `end program` statement is
```
Comment (valeriaRaffuzzi) [inline UserInterface/fileOutput/outputFile_class.f90:277]: This comment is not clear.

File: UserInterface/fileOutput/outputFile_class.f90 (hunk 4 above)
Comment (Mikolaj-A-Kowalski) [reply UserInterface/fileOutput/outputFile_class.f90:240]: :fearful: Ups... not great thing to forget

File: UserInterface/fileOutput/dummyPrinter_class.f90 (hunk 3 above)
Comment (Mikolaj-A-Kowalski) [reply UserInterface/fileOutput/dummyPrinter_class.f90:28]: It is... I think this is an oversight that this got left in.

File: UserInterface/fileOutput/outputFile_class.f90 (hunk 5 above)
Comment (Mikolaj-A-Kowalski) [reply UserInterface/fileOutput/outputFile_class.f90:277]:
```Fortran
  !!  NOTE:
  !!    Being marked `final`(a deconstructor)  this subroutine will be normally implicitly called by
  !!    by the compiler whenever a variable of `type(outputFile)` goes out of scope (e.g. when program 
  !!    execution reaches `end subroutine`, `end function` or `end block` statement. However, the rule in 
  !!    Fortran is that the decostructors are NOT called at the end of the program `end program`. In that case 
  !!    if one uses the output file as a module variable or defines it inside the program block, the `finalisation` 
  !!    must be called explicitly to ensure or data from the buffer is written to the disk 
  !!     
```
Does this work better?

**Review by valeriaRaffuzzi — CHANGES_REQUESTED** [review]
See comments.

File: UserInterface/fileOutput/dummyPrinter_class.f90 (hunk 3 above)
Comment (Mikolaj-A-Kowalski) [reply UserInterface/fileOutput/dummyPrinter_class.f90]: Should be removed now.

File: UserInterface/fileOutput/outputFile_class.f90 (hunk 4 above)
Comment (Mikolaj-A-Kowalski) [reply UserInterface/fileOutput/outputFile_class.f90:240]: Should be fixed.

File: UserInterface/fileOutput/outputFile_class.f90 (hunk 5 above)
Comment (Mikolaj-A-Kowalski) [reply UserInterface/fileOutput/outputFile_class.f90]: Comment has been changed.

**Comment by Mikolaj-A-Kowalski** [comment]
So I think I have addressed all the issues now... 
If it looks ok I will rebase on the current main (and squash all the fixups!) and then it should be ready to be merged.

File: docs/Output.rst [hunk 6]
```diff
@@ -76,9 +77,10 @@ Writing to an output file in SCONE is done through a series of calls to appropri
   !! Same goes for blocks
   call out % endBlock()
 
-  !! When done, we can write output to a file
-  !! Extension will be determined by the format printer
-  call out % writeToFile("./path/to/output/without_any_extension")
+  !! We need to make sure that output file is properly finalised. We can skip
+  !! this line if we we are not in a `program` block e.g. in `subroutine` or `function`
```
Comment (valeriaRaffuzzi) [inline docs/Output.rst:81]:
```suggestion
  !! this line if we are not in a `program` block e.g. in `subroutine` or `function`
```
Typo!

**Review by valeriaRaffuzzi — APPROVED** [review]
Very nice, I think this all looks great and it's a much better alternative to write the output file than keep it all in memory! 

I think there's just a small leftover typo that I spotted in the docs (soon SCONE will be typo free!!!) but then feel free to rebase and merge
